src/Oscoin/P2P/Disco/GCE.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DataKinds #-}
+
+module Oscoin.P2P.Disco.GCE where
+
+import           Oscoin.Prelude
+
+import           Oscoin.P2P.Types (Network, renderNetwork)
+
+import           Control.Monad.IO.Unlift (MonadUnliftIO)
+import           Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as Map
+import           Data.IP (IP)
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import           Data.Text.Read as T
+import           Lens.Micro
+                 ( Getting
+                 , each
+                 , filtered
+                 , (^?)
+                 , set
+                 , to
+                 , toListOf
+                 , traversed
+                 , _Just
+                 )
+import           Lens.Micro.Extras (view)
+import           Network.Google
+                 (AllowScopes, Env, HasScope, runGoogle, runResourceT, send)
+import           Network.Google.Compute
+                 ( Instance
+                 , InstanceAggregatedList
+                 , InstancesAggregatedList
+                 , NetworkInterface
+                 , i1Labels
+                 , i1NetworkInterfaces
+                 , ialItems
+                 , ialaFilter
+                 , ialiAddtional
+                 , ilAddtional
+                 , instancesAggregatedList
+                 , islInstances
+                 , niNetwork
+                 , niNetworkIP
+                 )
+import           Network.Socket (PortNumber)
+
+-- | Look up GCE instances as gossip peers.
+--
+-- The instances are filtered by the given labels. Each instance must be on the
+-- VPC denoted by 'Network', which is determined by:
+--
+-- * filtering by a "network" label with a value of 'Network'
+-- * considering only the NIC(s) of instances with a matching 'niNetwork'
+--
+-- Instances may advertise their bound ports via labels.
+--
+lookup
+    :: ( MonadUnliftIO m
+       , AllowScopes s
+       , HasScope    s InstancesAggregatedList
+       )
+    => Text              -- ^ Project ID
+    -> HashMap Text Text -- ^ Labels
+    -> Text              -- ^ Label to consider for obtaining 'PortNumber'
+    -> Network
+    -> Env s
+    -> m (Set (IP, Maybe PortNumber))
+lookup proj labels portLabel (renderNetwork -> net) r =
+    runResourceT . runGoogle r $ do
+        instances <- toListOf inst <$> send (mkRq proj)
+        pure . Set.fromList . flip concatMap instances $ \i ->
+            let
+                iport = firstOf  (port . traversed) i
+                addrs = toListOf addr               i
+             in
+                zip addrs (repeat iport)
+  where
+    -- morally. microlens is just .. too micro
```

Comment:
Don't even have proper prisms :trollface:

Reply:
Yeah what's the point really when you can't talk about a cocartesian coproduct?